File: test_python/problematic_code.py
#!/usr/bin/env python3
# Problematic Python code for testing linters


def bad_function(x, y, z):
    """This function has many issues"""
    if x is None:
        return
    
    # Long line that exceeds PEP 8 recommendations and should be split into multiple lines for better readability
    result = (
        x + y + z + 1 + 2 + 3 + 4 + 5 + 6 + 7 +
        8 + 9 + 10 + 11 + 12 + 13 + 14 + 15
    )
    
    # Missing type hints
    def inner_function(a, b):
        return a + b
    
    # Bare except
    try:
        result = inner_function(x, y)
    except Exception:
        pass
    
    # Mutable default argument
    def another_bad_function(items=[]):
        items.append("bad")
        return items
    
    # Missing return type annotation
    return result



class badClass:
    """Class with naming issues"""
    
    def __init__(self, name):
        self.name = name
        
    def method_with_issues(self,x,y):
        # No docstring
        # Comparison with None using ==
        if self.name is None:
            return False
            
        # String formatting issues
        _ = "Hello %s, you have %d items" % (self.name, x)
        
        return True


# Global variable (not recommended)
GLOBAL_VAR = "bad practice"




# Function without type hints
def process_data(data):
    if data is None:
        return []
    
    # List comprehension that could be simplified
    result = []
    for item in data:
        if item is not None:
            result.append(item.upper())
    
    return result

# Missing main guard
print("This should be in a main guard")
bad_function(1, 2, 3)

File: test_nodejs/problematic_code.js
// Problematic Node.js code for testing linters

var fs = require('fs');
var path = require('path');
var http = require('http');
var express = require('express');

function badFunction(x,y,z) {
    if (x == null) {
        return;
    }
    
    var unused_var = "this is not used";
    
    // Long line that exceeds recommended length and should be split into multiple lines for better readability and maintainability
    var result = x + y + z + 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 + 10 + 11 + 12 + 13 + 14 + 15 + 16 + 17 + 18 + 19 + 20;
    
    // Missing semicolon
    var data = {key: "value", another_key:"another_value"}
    
    // Bad comparison
    if (result == undefined) {
        console.log("result is undefined")
    }
    
    // Inconsistent quotes
    var message1 = "Hello world";
    var message2 = 'Hello world';
    
    return result
}

class BadClass {
    constructor(name,age) {
        this.name=name;
        this.age=age;
    }
    
    methodWithIssues(x,y) {
        // No var/let/const
        globalVar = "bad practice";
        
        // Eval usage (dangerous)
        eval("console.log('bad practice')");
        
        // == instead of ===
        if (this.name == null) {
            return false;
        }
        
        // Missing return
        var result = x + y;
    }
}

// Function without proper error handling
function processData(data) {
    if (data == null) {
        return [];
    }
    
    try {
        var parsed = JSON.parse(data);
    } catch (e) {
        // Empty catch block
    }
    
    return parsed;
}

// Callback hell
function nestedCallbacks(callback) {
    fs.readFile('file1.txt', function(err, data1) {
        if (err) throw err;
        fs.readFile('file2.txt', function(err, data2) {
            if (err) throw err;
            fs.readFile('file3.txt', function(err, data3) {
                if (err) throw err;
                callback(data1 + data2 + data3);
            });
        });
    });
}

// Missing 'use strict'
function oldStyleFunction() {
    arguments.callee; // Deprecated
    with (Math) { // Deprecated
        var x = cos(PI);
    }
}

// Inconsistent spacing and formatting
var obj={key1:"value1",key2:"value2"};
var arr=[1,2,3,4,5];

// Missing error handling in async operations
http.get('http://example.com', function(res) {
    res.on('data', function(chunk) {
        console.log(chunk);
    });
});

// Global variables
var GLOBAL_VAR = "bad practice";
window.globalVar = "also bad";

// Console.log left in code
console.log("Debug message that should be removed");

// Unreachable code
function unreachableCode() {
    return true;
    console.log("This will never execute");
}

// Missing JSDoc
function complexFunction(param1, param2, param3) {
    return param1 * param2 + param3;
}

// Prototype pollution vulnerability
Object.prototype.isAdmin = true;

// Using deprecated Buffer constructor
var buffer = new Buffer('hello');

// Missing await
async function asyncFunction() {
    Promise.resolve('hello');
    return 'done';
}

// Export at the end (should be at top)
module.exports = {
    badFunction: badFunction,
    BadClass: BadClass,
    processData: processData
};
